File: Sorting/57_Insert_Interval/solution57.cpp
#include <vector>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int>> insert(vector<vector<int>>& intervals, vector<int>& newInterval) {
        vector<vector<int>> res;
        
        for (int i = 0; i < intervals.size(); ++i) {
            if (newInterval[1] < intervals[i][0]) {
                res.push_back(newInterval);
                res.insert(res.end(), intervals.begin() + i, intervals.end());
                return res;
            }
            else if (newInterval[0] > intervals[i][1]) {
                res.push_back(intervals[i]);
            }
            else {
                newInterval[0] = min(newInterval[0], intervals[i][0]);
                newInterval[1] = max(newInterval[1], intervals[i][1]);
            }
        }
        
        res.push_back(newInterval);
        return res;
    }
};

// Runtime1: 8ms, beats 80.37%
// Memory1: 20.71MB, beats 74.69%

// Runtime2: 7ms, beats 91.68%
// Memory2: 20.54MB, beats 98.70%

// Runtime3: 7ms, beats 91.68%
// Memory3: 20.80MB, beats 38.21%

File: Sorting/57_Insert_Interval/solution57.py
class Solution:
    def insert(self, intervals: List[List[int]], newInterval: List[int]) -> List[List[int]]:
        res = []

        for i in range(len(intervals)):
            if newInterval[1] < intervals[i][0]:
                res.append(newInterval)
                return res + intervals[i:]
            elif newInterval[0] > intervals[i][1]:
                res.append(intervals[i])
            else:
                newInterval = [
                    min(newInterval[0], intervals[i][0]),
                    max(newInterval[1], intervals[i][1]),
                ]
        res.append(newInterval)
        return res

# Runtime1: 70ms, beats 83.71%
# Memory1: 19.96MB, beats 19.68%

# Runtime2: 68ms, beats 89.76%
# Memory2: 19.97MB, beats 19.68%

# Runtime3: 68ms, beats 89.76%
# Memory3: 19.90MB, beats 77.20%